

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Getting Started &mdash; smcsmc 0.1beta documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script type="text/javascript" src="_static/jquery.js"></script>
        <script type="text/javascript" src="_static/underscore.js"></script>
        <script type="text/javascript" src="_static/doctools.js"></script>
        <script type="text/javascript" src="_static/language_data.js"></script>
        <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="General Usage" href="usage/usage.html" />
    <link rel="prev" title="Introduction" href="introduction.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> smcsmc
          

          
          </a>

          
            
            
              <div class="version">
                latest
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="introduction.html">Introduction</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Getting Started</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#basic-concepts">Basic Concepts</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#input-format">Input Format</a></li>
<li class="toctree-l3"><a class="reference internal" href="#basic-arguements">Basic Arguements</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#running-smcsmc">Running <code class="code docutils literal notranslate"><span class="pre">smcsmc</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#output">Output</a></li>
<li class="toctree-l2"><a class="reference internal" href="#visualising-output">Visualising Output</a><ul class="simple">
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="usage/usage.html">General Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorials/tutorials.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="api.html">API Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="cli.html">Command Line Interface</a></li>
<li class="toctree-l1"><a class="reference internal" href="glossary.html">Glossary</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">smcsmc</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>Getting Started</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/getting_started.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="getting-started">
<span id="id1"></span><h1>Getting Started<a class="headerlink" href="#getting-started" title="Permalink to this headline">¶</a></h1>
<p>This guide will familiarise your with the concepts necessary to run analyses in SMCSMC and take you through the basic aspects of an analysis using  toy data. You may also be interested in one of several examples using real and simulated data in the <a class="reference internal" href="tutorials/tutorials.html#tutorial"><span class="std std-ref">Tutorials</span></a> section.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This tutorial is intended to be run on a personal computer, and values have been scaled accordingly. For analysis on real data, we expect that a user will have access to a compute cluster, and we provide guidance about setting up <code class="code docutils literal notranslate"><span class="pre">smcsmc</span></code> to function with common architechtures in <a class="reference internal" href="usage/cluster.html#cluster"><span class="std std-ref">Clustering Computing</span></a>.</p>
</div>
<div class="section" id="basic-concepts">
<h2>Basic Concepts<a class="headerlink" href="#basic-concepts" title="Permalink to this headline">¶</a></h2>
<p>SMCSMC is a <a class="reference internal" href="glossary.html#term-particle-filter"><span class="xref std std-term">particle filter</span></a>, which means that a given number of <a class="reference internal" href="glossary.html#term-particle"><span class="xref std std-term">particles</span></a> are simulated, evaluated for an approximate likelihood, and resampled until demographic parameters have converged. Many of the options relate to the behaviour of this particle filter, and the remainder deal with the demographic model that you wish to infer. SMCSMC can infer complex demographic models involving several populations, though you should be weary of overspecifying your model. Run time increases drastically with the number of parameters needed to specify the model.</p>
<div class="section" id="input-format">
<h3>Input Format<a class="headerlink" href="#input-format" title="Permalink to this headline">¶</a></h3>
<p>Suppose we have the following data, saved into a file named <code class="code docutils literal notranslate"><span class="pre">toy.seg</span></code>.</p>
<table class="colwidths-given docutils align-default">
<colgroup>
<col style="width: 17%" />
<col style="width: 17%" />
<col style="width: 17%" />
<col style="width: 17%" />
<col style="width: 17%" />
<col style="width: 17%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>segment start</p></th>
<th class="head"><p>segment length</p></th>
<th class="head"><p>invariant</p></th>
<th class="head"><p>break</p></th>
<th class="head"><p>chr</p></th>
<th class="head"><p>genotype</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>1</p></td>
<td><p>521</p></td>
<td><p>T</p></td>
<td><p>F</p></td>
<td><p>1</p></td>
<td><p>01.0</p></td>
</tr>
<tr class="row-odd"><td><p>522</p></td>
<td><p>2721</p></td>
<td><p>T</p></td>
<td><p>F</p></td>
<td><p>1</p></td>
<td><p>0111</p></td>
</tr>
<tr class="row-even"><td><p>3243</p></td>
<td><p>1758</p></td>
<td><p>T</p></td>
<td><p>F</p></td>
<td><p>1</p></td>
<td><p>10//</p></td>
</tr>
<tr class="row-odd"><td><p>5001</p></td>
<td><p>1296</p></td>
<td><p>T</p></td>
<td><p>F</p></td>
<td><p>1</p></td>
<td><p>0000</p></td>
</tr>
<tr class="row-even"><td><p>6297</p></td>
<td><p>1</p></td>
<td><p>T</p></td>
<td><p>F</p></td>
<td><p>1</p></td>
<td><p>….</p></td>
</tr>
<tr class="row-odd"><td><p>6298</p></td>
<td><p>4669</p></td>
<td><p>T</p></td>
<td><p>F</p></td>
<td><p>1</p></td>
<td><p>0110</p></td>
</tr>
<tr class="row-even"><td><p>10967</p></td>
<td><p>880</p></td>
<td><p>T</p></td>
<td><p>T</p></td>
<td><p>1</p></td>
<td><p>0100</p></td>
</tr>
<tr class="row-odd"><td><p>1</p></td>
<td><p>708</p></td>
<td><p>T</p></td>
<td><p>F</p></td>
<td><p>2</p></td>
<td><p>1010</p></td>
</tr>
</tbody>
</table>
<p>This gives the state of four haplotypes along the sequence, indicating segment starts and their lengths. If a segment is the last in a block before coordinates reset (i.e. when including multiple chromosomes in a single <code class="code docutils literal notranslate"><span class="pre">seg</span></code> file, it is called a genetic break. Genotypes in terms of either major/minor or ancestral/derived are given for each haplotype. Missing data is coded with a period (<code class="code docutils literal notranslate"><span class="pre">.</span></code>) whilst unphased variants are given with a forward slash (<code class="code docutils literal notranslate"><span class="pre">/</span></code>).</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This data format is similar to the input for <a class="reference external" href="https://github.com/stschiff/msmc">msmc</a> however, in the second column we require the length from the current segment to the next, rather than the number of called sites from the previous segment to the current one.</p>
</div>
<p>Segment files are specified by the <code class="code docutils literal notranslate"><span class="pre">seg</span></code> flag, or if more than one are used, by the <code class="code docutils literal notranslate"><span class="pre">segs</span></code> flag. Globs are encouraged for readability. In the case of multiple files, genetic breaks are inferred from the beginnings and ends of files. Additionally, we tell <code class="code docutils literal notranslate"><span class="pre">smcsmc</span></code> the number of samples we expect with <code class="code docutils literal notranslate"><span class="pre">nsam</span></code>, so it knows to check the <code class="code docutils literal notranslate"><span class="pre">seg</span></code> file for formatting issues.</p>
</div>
<div class="section" id="basic-arguements">
<h3>Basic Arguements<a class="headerlink" href="#basic-arguements" title="Permalink to this headline">¶</a></h3>
<p>In addition to the segment file, we need a few more pieces of information to kick off the particle filter.</p>
<p><strong>Particle Count:</strong> We must specify a particle count (<code class="code docutils literal notranslate"><span class="pre">Np</span></code>) or use the default value of 1000. The number of particles refers to the number of individual ARGs which are simulated by <code class="code docutils literal notranslate"><span class="pre">SCRM</span></code>, higher numbers means that the model is more likely to converge on a reasonable answer, but increasing the particle count is computationally expensive. We recommend particle counts between 25 and 50 thousand for analysis. However, for exploratory work, smaller particle counts (a good starting point is 10 thousand) can be used to more effectively use computational resources.  In this guide, intended for use on a personal computer, we will use 10 particles.</p>
<p><strong>Number of Iterations:</strong> We also explictly define the number of expectation maximization iterations that we want to use (<code class="code docutils literal notranslate"><span class="pre">EM</span></code>). A good place to start is 15 iterations. <code class="code docutils literal notranslate"><span class="pre">smcsmc</span></code> looks for output before starting, so should you be unsatisfied with the convergence after 15 iterations, simply run the exact same program with a higher number of iterations. <code class="code docutils literal notranslate"><span class="pre">smcsmc</span></code> will find the output for the iterations already run and simply continue from where it left off.</p>
<p><strong>Demographic Parameters:</strong> Several arguements are good to specify, especially when analysing real data as they help with convergence. Here we will use a mutation rate (<code class="code docutils literal notranslate"><span class="pre">mu</span></code>) of <code class="code docutils literal notranslate"><span class="pre">1.25e-8</span></code> and a recombination rate (<code class="code docutils literal notranslate"><span class="pre">rho</span></code>) of <code class="code docutils literal notranslate"><span class="pre">3e-9</span></code>. We give an effective population size (<code class="code docutils literal notranslate"><span class="pre">N0</span></code>) of 14312, and infer trees back to <code class="code docutils literal notranslate"><span class="pre">4*N0</span></code> generations with <code class="code docutils literal notranslate"><span class="pre">tmax</span></code>. We set this to 3.5, which is 1.2 Mya with an <code class="code docutils literal notranslate"><span class="pre">N0</span></code> of 10000. <code class="code docutils literal notranslate"><span class="pre">smcsmc</span></code> infers demographic parameters as discrete over intervals. We specify 31 intervals evenly spaced on the log scale with <code class="code docutils literal notranslate"><span class="pre">P</span> <span class="pre">133</span> <span class="pre">133032</span> <span class="pre">31*1</span></code> giving the end times of the first and last epochs in generations, and the pattern for their generation.</p>
<p>We also give the path to the output folder with <code class="code docutils literal notranslate"><span class="pre">o</span></code>.</p>
</div>
</div>
<div class="section" id="running-smcsmc">
<h2>Running <code class="code docutils literal notranslate"><span class="pre">smcsmc</span></code><a class="headerlink" href="#running-smcsmc" title="Permalink to this headline">¶</a></h2>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>It is always a good idea to start a new <code class="code docutils literal notranslate"><span class="pre">conda</span></code> environment for a new analysis to ensure that there are no dependecy conflicts, though you can skip this step if you certain that your environment is correctly configured.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>conda create --name smcsmc_tutorial
conda activate smcsmc_tutorial
</pre></div>
</div>
</div>
<p>Once <code class="code docutils literal notranslate"><span class="pre">smcsmc</span></code> is installed, we can format the arguements detailed above into a dictionary.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">args</span> <span class="o">=</span> <span class="p">{</span>
     <span class="s1">&#39;seg&#39;</span><span class="p">:</span>                  <span class="s1">&#39;test.seg&#39;</span><span class="p">,</span>
     <span class="s1">&#39;nsam&#39;</span><span class="p">:</span>                 <span class="s1">&#39;4&#39;</span><span class="p">,</span>
     <span class="s1">&#39;Np&#39;</span><span class="p">:</span>                   <span class="s1">&#39;10&#39;</span><span class="p">,</span>
     <span class="s1">&#39;EM&#39;</span><span class="p">:</span>                   <span class="s1">&#39;1&#39;</span><span class="p">,</span>
     <span class="s1">&#39;mu&#39;</span><span class="p">:</span>                   <span class="s1">&#39;1.25e-8&#39;</span><span class="p">,</span>
     <span class="s1">&#39;rho&#39;</span><span class="p">:</span>                  <span class="s1">&#39;rho&#39;</span><span class="p">,</span>
     <span class="s1">&#39;N0&#39;</span><span class="p">:</span>                   <span class="s1">&#39;10000&#39;</span><span class="p">,</span>
     <span class="s1">&#39;tmax&#39;</span><span class="p">:</span>                 <span class="s1">&#39;3.5&#39;</span>
     <span class="s1">&#39;P&#39;</span><span class="p">:</span>                    <span class="s1">&#39;133 133032 31*1&#39;</span>
     <span class="s1">&#39;no-infer-recomb&#39;</span><span class="p">:</span>      <span class="s1">&#39;&#39;</span><span class="p">,</span>
     <span class="s1">&#39;o&#39;</span><span class="p">:</span>                    <span class="s1">&#39;smcsmc_output&#39;</span>
<span class="p">}</span>
</pre></div>
</div>
<p>We directly use this dictionary with the <code class="code docutils literal notranslate"><span class="pre">run_smcsmc</span></code> command, which takes as its only arguement a dictionary of arguements.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">smcsmc</span>
<span class="n">smcsmc</span><span class="o">.</span><span class="n">run_smcsmc</span><span class="p">(</span><span class="n">args</span><span class="p">)</span>
</pre></div>
</div>
<p>If your installation has been successful, then this will begin the process of parsing the input, merging any given <code class="code docutils literal notranslate"><span class="pre">seg</span></code> files, starting the particle filter, and iterating through the <code class="code docutils literal notranslate"><span class="pre">EM</span></code> steps requested.</p>
</div>
<div class="section" id="output">
<h2>Output<a class="headerlink" href="#output" title="Permalink to this headline">¶</a></h2>
<p>If your <code class="code docutils literal notranslate"><span class="pre">smcsmc</span></code> has run correctly, the resulting output directory will look something like this, with a seperate folder for each EM iteration, and a seperate file for each chunk, if this option has been used.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>output/
        emiterN/
                chunkN.out
                chunkN.stdout
                chunk.stderr
                chunkfinal.out
        merged.seg
        merged.map
        result.log
        result.out
</pre></div>
</div>
<p>If you are following this tutorial and are only using a single input <code class="code docutils literal notranslate"><span class="pre">seg</span></code> file, you will not see <code class="code docutils literal notranslate"><span class="pre">merged.seg</span></code> or <code class="code docutils literal notranslate"><span class="pre">merged.map</span></code> as there was no need to generate them. The output for each chunk is given, along with stdout and sterr, and results are aggregated over all chunks each epoch into <code class="code docutils literal notranslate"><span class="pre">chunkfinal.out</span></code>. The final epoch will be post processed into <code class="code docutils literal notranslate"><span class="pre">result.out</span></code>. Output and debugging information along with useful information to help interpret the results of your model are given in <code class="code docutils literal notranslate"><span class="pre">result.log</span></code>.</p>
<p>An example of a <cite>results.out</cite> file is given here:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>Iter  Epoch       Start       End   Type   From     To            Opp          Count           Rate             Ne         ESS  Clump
<span class="m">15</span>      <span class="m">0</span>           <span class="m">0</span>         <span class="m">133</span>   Coal      <span class="m">0</span>     -1      <span class="m">307466</span>.77      <span class="m">18</span>.798649  <span class="m">6</span>.1140424e-05      <span class="m">8177</span>.8955      <span class="m">1</span>.0272     -1
<span class="m">15</span>      <span class="m">0</span>           <span class="m">0</span>         <span class="m">133</span>   Coal      <span class="m">1</span>     -1      <span class="m">324087</span>.66      <span class="m">55</span>.469939  <span class="m">0</span>.00017115721      <span class="m">2921</span>.2909      <span class="m">1</span>.0225     -1
<span class="m">15</span>      <span class="m">1</span>         <span class="m">133</span>       <span class="m">166</span>.2   Coal      <span class="m">0</span>     -1       <span class="m">167571</span>.8      <span class="m">1</span>.0002328  <span class="m">5</span>.9689802e-06      <span class="m">83766</span>.402      <span class="m">1</span>.0791     -1
<span class="m">15</span>      <span class="m">1</span>         <span class="m">133</span>       <span class="m">166</span>.2   Coal      <span class="m">1</span>     -1      <span class="m">174231</span>.23    <span class="m">0</span>.067992116  <span class="m">3</span>.9024069e-07      <span class="m">1281260</span>.5      <span class="m">1</span>.0674     -1
<span class="m">15</span>      <span class="m">2</span>       <span class="m">166</span>.2      <span class="m">207</span>.68   Coal      <span class="m">0</span>     -1      <span class="m">259801</span>.05      <span class="m">40</span>.479172  <span class="m">0</span>.00015580835      <span class="m">3209</span>.0707      <span class="m">1</span>.1686     -1
<span class="m">15</span>      <span class="m">2</span>       <span class="m">166</span>.2      <span class="m">207</span>.68   Coal      <span class="m">1</span>     -1      <span class="m">267803</span>.96  <span class="m">0</span>.00028150086  <span class="m">1</span>.0511452e-09  <span class="m">4</span>.7567166e+08      <span class="m">1</span>.1522     -1
<span class="m">15</span>     -1           <span class="m">0</span>       1e+99  Delay     -1     -1  <span class="m">2</span>.6001037e+09              <span class="m">0</span>              <span class="m">0</span>              <span class="m">0</span>           <span class="m">1</span>     -1
<span class="m">15</span>     -1           <span class="m">0</span>       1e+99   LogL     -1     -1              <span class="m">1</span>      -31005174      -31005174              <span class="m">0</span>           <span class="m">1</span>     -1
<span class="m">15</span>      <span class="m">0</span>           <span class="m">0</span>         <span class="m">133</span>   Migr      <span class="m">0</span>      <span class="m">1</span>       <span class="m">178112</span>.7      <span class="m">7</span>.1251662  <span class="m">4</span>.0003695e-05              <span class="m">0</span>      <span class="m">1</span>.0385     -1
<span class="m">15</span>      <span class="m">0</span>           <span class="m">0</span>         <span class="m">133</span>   Migr      <span class="m">1</span>      <span class="m">0</span>      <span class="m">186343</span>.02  <span class="m">0</span>.00022362676  <span class="m">1</span>.2000812e-09              <span class="m">0</span>      <span class="m">1</span>.0338     -1
<span class="m">15</span>      <span class="m">1</span>         <span class="m">133</span>       <span class="m">166</span>.2   Migr      <span class="m">0</span>      <span class="m">1</span>       <span class="m">87712</span>.37      <span class="m">0</span>.5934193  <span class="m">6</span>.7655144e-06              <span class="m">0</span>      <span class="m">1</span>.0819     -1
<span class="m">15</span>      <span class="m">1</span>         <span class="m">133</span>       <span class="m">166</span>.2   Migr      <span class="m">1</span>      <span class="m">0</span>       <span class="m">91253</span>.03      <span class="m">3</span>.0913275  <span class="m">3</span>.3876437e-05              <span class="m">0</span>      <span class="m">1</span>.0713     -1
<span class="m">15</span>      <span class="m">2</span>       <span class="m">166</span>.2      <span class="m">207</span>.68   Migr      <span class="m">0</span>      <span class="m">1</span>      <span class="m">134755</span>.98      <span class="m">1</span>.0135379  <span class="m">7</span>.5212833e-06              <span class="m">0</span>      <span class="m">1</span>.1711     -1
<span class="m">15</span>      <span class="m">2</span>       <span class="m">166</span>.2      <span class="m">207</span>.68   Migr      <span class="m">1</span>      <span class="m">0</span>      <span class="m">139267</span>.34      <span class="m">6</span>.0189739  <span class="m">4</span>.3218847e-05              <span class="m">0</span>      <span class="m">1</span>.1565     -1
</pre></div>
</div>
<p>The full interpretation of the output is somewhat involved, and we recommend referring to the supporting publication for a complete explaination of all relevant quantities. However, in brief:</p>
<ul class="simple">
<li><p><strong>Iter</strong>: The EM iteration to which this file refers. In the case of aggregated files (<code class="code docutils literal notranslate"><span class="pre">chunkfinal.out</span></code>, <code class="code docutils literal notranslate"><span class="pre">result.out</span></code>), multiple iterations are represented in the same file, and it is often useful either to graph each iteration seperately to show convergence, or to select the last iteration to show the final inference.</p></li>
<li><p><strong>Epoch</strong>: The discrete piece of time in which rates are assumed to be constant. The pattern which epochs follow is set by the <code class="code docutils literal notranslate"><span class="pre">P</span></code> flag.</p></li>
<li><p><strong>Start</strong>: The beginning of the epoch, in generations.</p></li>
<li><p><strong>End</strong>: The end of the poch, in generations.</p></li>
<li><p><strong>Type</strong>: Either <code class="code docutils literal notranslate"><span class="pre">Coal</span></code> (coalescence), <code class="code docutils literal notranslate"><span class="pre">Delay</span></code>, <code class="code docutils literal notranslate"><span class="pre">LogL</span></code> (Log likelihood), or <code class="code docutils literal notranslate"><span class="pre">Migr</span></code>. The important lines for interpreting output are:</p>
<ul>
<li><p><code class="code docutils literal notranslate"><span class="pre">Coal</span></code>: This line refers to coalescent events. For <code class="code docutils literal notranslate"><span class="pre">Coal</span></code> events, the <code class="code docutils literal notranslate"><span class="pre">From</span></code> refers to the population of interest while the <code class="code docutils literal notranslate"><span class="pre">To</span></code> is meaningless. An estimate of <code class="code docutils literal notranslate"><span class="pre">Ne</span></code> is shown only in <code class="code docutils literal notranslate"><span class="pre">Coal</span></code> lines.</p></li>
<li><p><code class="code docutils literal notranslate"><span class="pre">Migr</span></code>: This line refers to migration events. <code class="code docutils literal notranslate"><span class="pre">From</span></code> and <code class="code docutils literal notranslate"><span class="pre">To</span></code> refer to the source and the sink of migration, backwards in time.</p></li>
</ul>
</li>
<li><p><strong>Opp</strong>: The possible time where events of this can may happen in this sequence. For recombination inference, this refers to the amount of time multiplied by the amount of sequence.</p></li>
<li><p><strong>Count</strong>: The number of events of this type inferred in this epoch.</p></li>
<li><p><strong>Rate</strong>: Either the coalescent, recombination, or migration rate. This is equal to the <code class="code docutils literal notranslate"><span class="pre">Count</span></code> divided by <code class="code docutils literal notranslate"><span class="pre">Opp</span></code>. Coalescent rates are number of coalescentn events per generation. Migration rates represent proportion migration from the source (<code class="code docutils literal notranslate"><span class="pre">From</span></code>) to the sink (<code class="code docutils literal notranslate"><span class="pre">To</span></code>) <strong>backwards in time</strong>.</p></li>
</ul>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p><code class="code docutils literal notranslate"><span class="pre">SCRM</span></code> reports migration rates <strong>forwards in time</strong> while <code class="code docutils literal notranslate"><span class="pre">smcsmc</span></code> reports it <strong>backwards in time</strong>.</p>
</div>
<ul class="simple">
<li><p><strong>Ne</strong>: The estimated population size in this epoch. Defined to be <code class="code docutils literal notranslate"><span class="pre">Opp</span></code> divded by <code class="code docutils literal notranslate"><span class="pre">2*Count</span></code>.</p></li>
<li><p><strong>ESS</strong>: Effective sample size, which is an indicator of the diversity of our particles. This is used when deciding to resample, or when to spawn new particles.</p></li>
<li><p><strong>Clump</strong>: This is defined in aggregated output, and not defined in intermediary files.</p></li>
</ul>
<p>Important diagnostic information about the reliability of inferred rates is contained in the results file. For instance, it is often useful to look at the amount of events inferred for each epoch. In general, it should monotonically increase, though this will depend on the definition of your epochs.  Can you spot the oddity above?</p>
<p>Very low counts in a particular epoch can be due to any number of reasons, but a good first step is to increase the number of particles.</p>
</div>
<div class="section" id="visualising-output">
<h2>Visualising Output<a class="headerlink" href="#visualising-output" title="Permalink to this headline">¶</a></h2>
<p>We provide several functions for visualising different aspects of the output.</p>
<p>It is a good idea to first check the convergence of your model by plotting each iteration seperately in what we refer to as a rainbow plot. Specify the path to your output <code class="code docutils literal notranslate"><span class="pre">result.out</span></code> file and the path to a preferred file path to store the image.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">result_path</span> <span class="o">=</span> <span class="s1">&#39;output/result.out&#39;</span>
<span class="n">plot_path</span> <span class="o">=</span> <span class="s1">&#39;output/rainbow.png&#39;</span>

<span class="n">smcsmc</span><span class="o">.</span><span class="n">plot_migration</span><span class="p">(</span><span class="n">result_path</span><span class="p">,</span> <span class="n">plot_path</span><span class="p">)</span>
</pre></div>
</div>
<p>An example plot is shown below, note that we have additionally specified a model from which this data was simulated. For details, please refer to the tutorials.</p>
<div class="figure align-center" id="id2">
<img alt="_images/rainbow.png" src="_images/rainbow.png" />
<p class="caption"><span class="caption-text">Estimated population size over 15 EM iterations each plotted in a different colour. The model appears to have well converged to a solution.</span><a class="headerlink" href="#id2" title="Permalink to this image">¶</a></p>
</div>
<p>Plotting migration and effective population size can be done similarly with <code class="code docutils literal notranslate"><span class="pre">smcsmc.plot_migration</span></code> and <cite>smcsmc.plot_ne</cite>. See the API documentation for more information.</p>
<div class="toctree-wrapper compound">
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="usage/usage.html" class="btn btn-neutral float-right" title="General Usage" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="introduction.html" class="btn btn-neutral float-left" title="Introduction" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, Christopher B Cole, Donna Henderson, Sha (Joe) Zhu, Gerton Lunter

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>